File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let prevMax = nums[0]
    let prevMin = nums[0]
    let max = nums[0]

    for(let i=1; i<nums.length; i++){
        const currMax = Math.max(prevMax * nums[i], nums[i], prevMin * nums[i])
        const currMin = Math.min(prevMax * nums[i], nums[i], prevMin * nums[i])

        prevMax = currMax
        prevMin = currMin

        max = Math.max(max, prevMax)
    }
    return max

};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_d = nums[0]
        min_d = nums[0]
        res = nums[0]

        for num in nums[1:]:
            if num < 0:
                max_d, min_d = min_d, max_d
            max_d = max(num, max_d * num)
            min_d = min(num, min_d * num)
            res = max(res, max_d)
        
        return res